pkg/controller/volume/persistentvolume/scheduler_binder.go

```diff
@@ -249,66 +252,157 @@ func (b *volumeBinder) AssumePodVolumes(assumedPod *v1.Pod, nodeName string) (al
 		newProvisionedPVCs = append(newProvisionedPVCs, claimClone)
 	}
 
-	if len(newProvisionedPVCs) != 0 {
-		bindingRequired = true
+	// Update cache with the assumed pvcs and pvs
+	if len(newBindings) > 0 {
+		b.podBindingCache.UpdateBindings(assumedPod, nodeName, newBindings)
+	}
+	if len(newProvisionedPVCs) > 0 {
 		b.podBindingCache.UpdateProvisionedPVCs(assumedPod, nodeName, newProvisionedPVCs)
 	}
 
 	return
 }
 
-// BindPodVolumes gets the cached bindings and PVCs to provision in podBindingCache
-// and makes the API update for those PVs/PVCs.
+// BindPodVolumes gets the cached bindings and PVCs to provision in podBindingCache,
+// makes the API update for those PVs/PVCs, and waits for the PVCs to be completely bound
+// by the PV controller.
 func (b *volumeBinder) BindPodVolumes(assumedPod *v1.Pod) error {
 	podName := getPodName(assumedPod)
-	glog.V(4).Infof("BindPodVolumes for pod %q", podName)
+	glog.V(4).Infof("BindPodVolumes for pod %q, node %q", podName, assumedPod.Spec.NodeName)
 
 	bindings := b.podBindingCache.GetBindings(assumedPod, assumedPod.Spec.NodeName)
 	claimsToProvision := b.podBindingCache.GetProvisionedPVCs(assumedPod, assumedPod.Spec.NodeName)
 
+	// Start API operations
+	err := b.bindAPIUpdate(podName, bindings, claimsToProvision)
+	if err != nil {
+		return err
+	}
+
+	return wait.Poll(time.Second, b.bindTimeout, func() (bool, error) {
```

Comment:
I don't know how computationally expensive `checkBindings` is. If it is expensive, running this every second may hinder Scheduler's performance. There are a couple of alternatives: 1) when a volume is bound, the status is updated on the API server. The volumeBinder watches those events and runs checkBindings only at those events. For example, BindPodVolumes runs checkBindings and calls sync.Cond.Wait when checkBinding returns false. The event handler that watches volume binding calls sync.Cond.Broadcast when a volume is bound. 2) Run this less frequently (this is obviously an inferior solution).

Reply:
`checkBindings` will run through all the PVCs in a single Pod, get the cached PVC and PV objects from a map, and check some fields on those objects.  I would expect Pods to have <5 persistent volumes most of the time, so I would expect the most expensive work to be the cache lookups (map).

I will look into the event handling alternative.  One potential concern with a Broadcast is all the threads will wake up at the same time so there could be many spikes.